### PR TITLE
Develop

### PR DESCRIPTION
**Pull request by ArtemKrotenok**

### CODE REVIEW

File: web-module/src/main/java/com/epam/esm/app/controller/api/GiftCertificateController.java [hunk 1]
```diff
@@ -0,0 +1,147 @@
...
+import lombok.extern.slf4j.Slf4j;
+import org.springframework.http.HttpStatus;
+import org.springframework.http.MediaType;
+import org.springframework.http.ResponseEntity;
+import org.springframework.web.bind.annotation.*;
+
+@RestController
+@RequestMapping(value = "/api/certificate", produces = MediaType.APPLICATION_JSON_VALUE)
+@AllArgsConstructor
+@Slf4j
```
Comment (andrey7067) [inline web-module/src/main/java/com/epam/esm/app/controller/api/GiftCertificateController.java:16]: You added a logger, but you don't use it

File: web-module/src/main/java/com/epam/esm/app/controller/api/GiftCertificateController.java [hunk 2]
```diff
@@ -0,0 +1,147 @@
...
+
+/**
+ * controller class for <b>gift certificate</b>.
+ * @autor Artem Krotenok
+ * @version 1.0
+ */
```
Comment (andrey7067) [inline web-module/src/main/java/com/epam/esm/app/controller/api/GiftCertificateController.java:22]: Move the documentation up

File: web-module/src/main/java/com/epam/esm/app/controller/api/GiftCertificateController.java [hunk 3]
```diff
@@ -0,0 +1,147 @@
...
+    public ResponseEntity<Object> createGiftCertificate(@RequestBody GiftCertificateDTO giftCertificateDTO) {
+        try {
+            return ResponseEntity
+                    .status(HttpStatus.CREATED)
+                    .body(giftCertificateService.create(giftCertificateDTO));
+        } catch (GiftCertificateServiceException exception) {
+            return ResponseEntity
+                    .status(exception.getResponseHttpStatus())
+                    .body(exception.getErrorResponseDTO());
+        }
```
Comment (andrey7067) [inline web-module/src/main/java/com/epam/esm/app/controller/api/GiftCertificateController.java:44]: Try to rewrite

**Review by andrey7067 — CHANGES_REQUESTED** [review]

File: repository-module/src/main/java/com/epam/esm/repository/impl/GiftCertificateRepositoryImpl.java [hunk 4]
```diff
@@ -0,0 +1,150 @@
...
+    @Override
+    public List<GiftCertificate> findByTag(String tag) {
+        List<GiftCertificate> giftCertificateList = jdbcTemplate.query("SELECT DISTINCT " +
+                        "gc.id, gc.name, gc.description, gc.price, gc.duration, gc.create_date, gc.last_update_date " +
+                        "FROM gift_certificate AS gc " +
+                        "JOIN gift_certificate_tag AS gct ON gc.id = gct.id_gift_certificate " +
+                        "JOIN tag t ON gct.id_tag = t.id WHERE t.name = ? ORDER BY name ASC",
+                new BeanPropertyRowMapper<>(GiftCertificate.class),
+                tag);
+        if (giftCertificateList.size() > 0) {
```
[the remark this answers is not in the document]
Comment (ArtemKrotenok) [reply repository-module/src/main/java/com/epam/esm/repository/impl/GiftCertificateRepositoryImpl.java:102]: fixed

File: repository-module/src/main/java/com/epam/esm/repository/impl/GiftCertificateRepositoryImpl.java [hunk 5]
```diff
@@ -0,0 +1,150 @@
...
+    }
+
+    @Override
+    public GiftCertificate findById(Long id) {
+        GiftCertificate giftCertificate = jdbcTemplate.query("SELECT * FROM gift_certificate WHERE id=?",
+                        new BeanPropertyRowMapper<>(GiftCertificate.class),
+                        id)
+                .stream().findAny().orElse(null);
+        if (giftCertificate != null) {
+            giftCertificate.setTagList(getTagListForCertificate(giftCertificate));
```
[the remark this answers is not in the document]
Comment (ArtemKrotenok) [reply repository-module/src/main/java/com/epam/esm/repository/impl/GiftCertificateRepositoryImpl.java:38]: fixed

File: service-module/src/main/java/com/epam/esm/service/imp/GiftCertificateServiceImpl.java [hunk 6]
```diff
@@ -0,0 +1,219 @@
...
+            return ResponseDTOUtil.getSuccessResponseDTO(
+                    ResponseCode.CREATE);
+        }
+        throw new GiftCertificateServiceException(ResponseDTOUtil.getErrorResponseDTO(
+                ResponseCode.NOT_CREATE));
+    }
+
+    @Override
+    public List<GiftCertificateDTO> getAllByPageSorted(Integer page) throws GiftCertificateServiceException {
+        if (page == null || page < 0) {
```
[the remark this answers is not in the document]
Comment (ArtemKrotenok) [reply service-module/src/main/java/com/epam/esm/service/imp/GiftCertificateServiceImpl.java:51]: fixed

File: web-module/src/main/java/com/epam/esm/app/controller/api/GiftCertificateController.java (hunk 2 above)
Comment (ArtemKrotenok) [reply web-module/src/main/java/com/epam/esm/app/controller/api/GiftCertificateController.java:22]: fixed

File: service-module/src/main/java/com/epam/esm/service/util/GiftCertificateUtil.java [hunk 7]
```diff
@@ -0,0 +1,75 @@
...
+import java.sql.Timestamp;
+import java.text.DateFormat;
+import java.text.SimpleDateFormat;
+import java.time.LocalDateTime;
+import java.util.TimeZone;
+import java.util.stream.Collectors;
+
+public class GiftCertificateUtil {
+
+    public static final String DATE_PATTERN = "yyyy-MM-dd'T'HH:mm:ss.SSS";
```
[the remark this answers is not in the document]
Comment (ArtemKrotenok) [reply service-module/src/main/java/com/epam/esm/service/util/GiftCertificateUtil.java:16]: I don't need it, but the author of the task really needs it: "CreateDate, LastUpdateDate - format ISO 8601 (https://en.wikipedia.org/wiki/ISO_8601). Example: 2018-08-29T06:12:15.156."

File: service-module/src/main/java/com/epam/esm/service/imp/TagServiceImpl.java [hunk 8]
```diff
@@ -0,0 +1,124 @@
...
+            return ResponseDTOUtil.getSuccessResponseDTO(
+                    ResponseCode.DELETE);
+        }
+        throw new GiftCertificateServiceException(ResponseDTOUtil.getErrorResponseDTO(
+                ResponseCode.NOT_DELETE));
+    }
+
+    @Override
+    public List<TagDTO> getAllByPageSorted(Integer page) throws GiftCertificateServiceException {
+        if (page < 0) {
```
[the remark this answers is not in the document]
Comment (ArtemKrotenok) [reply service-module/src/main/java/com/epam/esm/service/imp/TagServiceImpl.java:86]: fixed

File: service-module/src/main/java/com/epam/esm/service/imp/TagServiceImpl.java [hunk 9]
```diff
@@ -0,0 +1,124 @@
...
+
+    private List<TagDTO> checkAndReturnResult(List<Tag> tagList) throws GiftCertificateServiceException {
+        if (!tagList.isEmpty()) {
+            return tagList.stream().map(TagUtil::convert).collect(Collectors.toList());
+        }
+        throw new GiftCertificateServiceException(ResponseDTOUtil.getErrorResponseDTO(
+                ResponseCode.NOT_FOUND));
+    }
+
+    private String getErrorValid(Long id) {
```
[the remark this answers is not in the document]
Comment (ArtemKrotenok) [reply service-module/src/main/java/com/epam/esm/service/imp/TagServiceImpl.java:108]: fixed

File: service-module/src/main/java/com/epam/esm/service/imp/GiftCertificateServiceImpl.java [hunk 10]
```diff
@@ -0,0 +1,219 @@
...
+import java.util.Comparator;
+import java.util.List;
+import java.util.stream.Collectors;
+
+@Service
+@Slf4j
+@AllArgsConstructor
+public class GiftCertificateServiceImpl implements GiftCertificateService {
+
+    private static final int FIST_PAGE = 1;
```
Comment (KristinaPotapovich) [inline service-module/src/main/java/com/epam/esm/service/imp/GiftCertificateServiceImpl.java:30]: Please correct the word "FIRST"

File: repository-module/src/main/java/com/epam/esm/repository/impl/GiftCertificateTagRepositoryImpl.java [hunk 11]
```diff
@@ -0,0 +1,91 @@
...
+    public List<GiftCertificateTag> findAll() {
+        return jdbcTemplate.query("SELECT * FROM gift_certificate_tag", new BeanPropertyRowMapper<>(GiftCertificateTag.class));
+    }
+
+    @Override
+    public GiftCertificateTag findById(Long id) {
+        return jdbcTemplate.query("SELECT * FROM gift_certificate_tag WHERE id=?",
+                        new BeanPropertyRowMapper<>(GiftCertificateTag.class),
+                        id)
+                .stream().findAny().orElse(null);
```
[the remark this answers is not in the document]
Comment (ArtemKrotenok) [reply repository-module/src/main/java/com/epam/esm/repository/impl/GiftCertificateTagRepositoryImpl.java:32]: fixed

File: service-module/src/main/java/com/epam/esm/service/util/GiftCertificateUtil.java (hunk 7 above)
[the remark this answers is not in the document]
Comment (ArtemKrotenok) [reply service-module/src/main/java/com/epam/esm/service/util/GiftCertificateUtil.java:16]: fixed

File: web-module/src/test/java/com/epam/esm/repository/TagRepositoryTest.java [hunk 12]
```diff
@@ -0,0 +1,131 @@
...
+    @Test
+    void getAllTagByPageSorted_returnSortedTags() {
+        List<Tag> tags = tagRepository.getAllByPageSorted(START_POSITION, ITEMS_BY_PAGE);
+        assertFalse(tags.isEmpty());
+        assertTrue(isSorted(tags));
+        assertEquals(tags.size(), ITEMS_BY_PAGE);
+    }
+
+    @Test
+    void findById_validId_tag() {
```
[the remark this answers is not in the document]
Comment (ArtemKrotenok) [reply web-module/src/test/java/com/epam/esm/repository/TagRepositoryTest.java:88]: fixed

File: web-module/src/main/java/com/epam/esm/app/controller/api/TagController.java [hunk 13]
```diff
@@ -0,0 +1,108 @@
...
+    public TagService tagService;
+
+    /**
+     * controller for create new tag
+     *
+     * @param tagDTO - object contain new tag model
+     * @return if all success return object SuccessResponseDTO, else return object ErrorResponseDTO
+     */
+    @PostMapping
+    public ResponseEntity<Object> creatTag(@RequestBody TagDTO tagDTO) {
```
[the remark this answers is not in the document]
Comment (ArtemKrotenok) [reply web-module/src/main/java/com/epam/esm/app/controller/api/TagController.java:35]: fixed

File: web-module/src/main/java/com/epam/esm/app/controller/api/TagController.java [hunk 14]
```diff
@@ -0,0 +1,108 @@
...
+import com.epam.esm.service.model.TagDTO;
+import lombok.AllArgsConstructor;
+import lombok.extern.slf4j.Slf4j;
+import org.springframework.http.HttpStatus;
+import org.springframework.http.MediaType;
+import org.springframework.http.ResponseEntity;
+import org.springframework.web.bind.annotation.*;
+
+@RestController
+@RequestMapping(value = "/api/certificate/tag", produces = MediaType.APPLICATION_JSON_VALUE)
```
[the remark this answers is not in the document]
Comment (ArtemKrotenok) [reply web-module/src/main/java/com/epam/esm/app/controller/api/TagController.java:14]: fixed

File: web-module/src/main/java/com/epam/esm/app/controller/api/GiftCertificateController.java [hunk 15]
```diff
@@ -0,0 +1,147 @@
...
+
+    /**
+     * controller for search a list of gift certificates by tag, name, description (list sorted by name gift certificates)
+     *
+     * @param tag         - one of tag for gift certificate (not required)
+     * @param name        - name for gift certificate (not required)
+     * @param description - description for gift certificate (not required)
+     * @return if all success return list GiftCertificateDTO, else return object ErrorResponseDTO
+     */
+    @GetMapping("/search")
```
[the remark this answers is not in the document]
Comment (ArtemKrotenok) [reply web-module/src/main/java/com/epam/esm/app/controller/api/GiftCertificateController.java:112]: fixed

[the remark this answers is not in the document]
Comment (ArtemKrotenok) [reply web-module/src/main/java/com/epam/esm/app/controller/api/GiftCertificateController.java:112]: + rewrote search logic

File: web-module/src/main/java/com/epam/esm/app/controller/api/GiftCertificateController.java [hunk 16]
```diff
@@ -0,0 +1,147 @@
...
+    }
+
+    /**
+     * controller for getting a list of gift certificates by page using pagination (list sorted by name gift certificates)
+     *
+     * @param page - number page for return
+     * @return if all success return list GiftCertificateDTO, else return object ErrorResponseDTO
+     */
+    @GetMapping
+    public ResponseEntity<Object> getGiftCertificateByPageSorted(@RequestParam(name = "page", required = false) Integer page) {
```
[the remark this answers is not in the document]
Comment (ArtemKrotenok) [reply web-module/src/main/java/com/epam/esm/app/controller/api/GiftCertificateController.java:73]: fixed

File: service-module/src/main/java/com/epam/esm/service/imp/GiftCertificateServiceImpl.java (hunk 10 above)
Comment (ArtemKrotenok) [reply service-module/src/main/java/com/epam/esm/service/imp/GiftCertificateServiceImpl.java:30]: fixed

File: service-module/src/main/java/com/epam/esm/service/imp/TagServiceImpl.java [hunk 17]
```diff
@@ -0,0 +1,124 @@
...
+        return checkAndReturnResult(tagList);
+    }
+
+    @Override
+    public List<TagDTO> getAllSorted() throws GiftCertificateServiceException {
+        List<Tag> tagList = tagRepository.getAllSorted();
+        return checkAndReturnResult(tagList);
+    }
+
+    private List<TagDTO> checkAndReturnResult(List<Tag> tagList) throws GiftCertificateServiceException {
```
[the remark this answers is not in the document]
Comment (ArtemKrotenok) [reply service-module/src/main/java/com/epam/esm/service/imp/TagServiceImpl.java:100]: fixed

File: service-module/src/test/java/com/epam/esm/service/GiftCertificateServiceTest.java [hunk 18]
```diff
@@ -0,0 +1,119 @@
...
+    @Test
+    void findById_validId_GiftCertificateDTO() throws GiftCertificateServiceException {
+        when(giftCertificateRepository.findById(GIFT_CERTIFICATE_TEST_ID)).thenReturn(TestServiceDataUtil.getValidGiftCertificate());
+        GiftCertificateDTO finedGiftCertificateDTO = giftCertificateService.findById(GIFT_CERTIFICATE_TEST_ID);
+        assertEquals(finedGiftCertificateDTO, TestServiceDataUtil.getValidGiftCertificateDTO());
+    }
+
+    @Test
+    void findById_invalidId_GiftCertificateServiceException() {
+        Exception exception = assertThrows(Exception.class, () -> {
```
[the remark this answers is not in the document]
Comment (ArtemKrotenok) [reply service-module/src/test/java/com/epam/esm/service/GiftCertificateServiceTest.java:70]: fixed

File: web-module/src/main/java/com/epam/esm/app/controller/api/GiftCertificateController.java (hunk 3 above)
Comment (ArtemKrotenok) [reply web-module/src/main/java/com/epam/esm/app/controller/api/GiftCertificateController.java:44]: fixed

File: web-module/src/main/java/com/epam/esm/app/controller/api/GiftCertificateController.java [hunk 19]
```diff
@@ -0,0 +1,147 @@
...
+import com.epam.esm.service.model.GiftCertificateDTO;
+import lombok.AllArgsConstructor;
+import lombok.extern.slf4j.Slf4j;
+import org.springframework.http.HttpStatus;
+import org.springframework.http.MediaType;
+import org.springframework.http.ResponseEntity;
+import org.springframework.web.bind.annotation.*;
+
+@RestController
+@RequestMapping(value = "/api/certificate", produces = MediaType.APPLICATION_JSON_VALUE)
```
[the remark this answers is not in the document]
Comment (ArtemKrotenok) [reply web-module/src/main/java/com/epam/esm/app/controller/api/GiftCertificateController.java:14]: fixed

File: web-module/src/main/java/com/epam/esm/app/controller/api/GiftCertificateController.java [hunk 20]
```diff
@@ -0,0 +1,147 @@
...
+import com.epam.esm.service.exception.GiftCertificateServiceException;
+import com.epam.esm.service.model.GiftCertificateDTO;
+import lombok.AllArgsConstructor;
+import lombok.extern.slf4j.Slf4j;
+import org.springframework.http.HttpStatus;
+import org.springframework.http.MediaType;
+import org.springframework.http.ResponseEntity;
+import org.springframework.web.bind.annotation.*;
+
+@RestController
```
[the remark this answers is not in the document]
Comment (ArtemKrotenok) [reply web-module/src/main/java/com/epam/esm/app/controller/api/GiftCertificateController.java:13]: fixed

File: service-module/src/main/java/com/epam/esm/service/imp/GiftCertificateServiceImpl.java [hunk 21]
```diff
@@ -0,0 +1,219 @@
...
+    public static final int RESULT_ONE_RECORD = 1;
+    private GiftCertificateRepository giftCertificateRepository;
+
+    @Override
+    public SuccessResponseDTO create(GiftCertificateDTO giftCertificateDTO) throws GiftCertificateServiceException {
+        String errorValidMessage = getErrorValidMessage(giftCertificateDTO);
+        if (errorValidMessage != null) {
```
[the remark this answers is not in the document]
Comment (ArtemKrotenok) [reply service-module/src/main/java/com/epam/esm/service/imp/GiftCertificateServiceImpl.java]: fixed

**Pull request merged**